lib/src/sink_base.dart

```diff
@@ -0,0 +1,168 @@
+/// called.
+abstract class EventSinkBase<T> implements EventSink<T> {
+  /// Whether [close] has been called and no more events may be written.
+  bool get _closed => _closeMemo.hasRun;
+
+  @override
+  void add(T data) {
+    _checkCanAddEvent();
+    onAdd(data);
+  }
+
+  /// A method that handles data events that are passed to the sink.
+  @visibleForOverriding
+  void onAdd(T data);
+
+  @override
+  void addError(Object error, [StackTrace? stackTrace]) {
+    _checkCanAddEvent();
+    onError(error, stackTrace);
+  }
+
+  /// A method that handles error events that are passed to the sink.
+  @visibleForOverriding
+  void onError(Object error, [StackTrace? stackTrace]);
+
+  @override
+  Future<void> close() => _closeMemo.runOnce(onClose);
+  final _closeMemo = AsyncMemoizer<void>();
+
+  /// A method that handles the sink being closed.
+  ///
+  /// This may return a future that completes once the stream sink has shut
+  /// down. If cleaning up can fail, the error may be reported in the returned
+  /// future.
+  @visibleForOverriding
+  FutureOr<void> onClose();
+
+  /// Asserts that the sink is in a state where adding an event is valid.
+  void _checkCanAddEvent() {
+    if (_closed) throw StateError('Cannot add event after closing');
+  }
+}
+
+/// An abstract class that implements [StreamSink] in terms of [onData],
+/// [onError], and [onClose] methods.
+///
+/// This takes care of ensuring that events can't be added after [close] is
+/// called or during a call to [onStream].
+abstract class StreamSinkBase<T> extends EventSinkBase<T>
+    implements StreamSink<T> {
+  /// Whether a call to [addStream] is ongoing.
+  bool _addingStream = false;
+
+  @override
+  Future<void> get done => _closeMemo.future;
+
+  @override
+  Future<void> addStream(Stream<T> stream) {
+    _checkCanAddEvent();
+
+    _addingStream = true;
+    var completer = Completer<void>.sync();
+    stream.listen(onAdd, onError: onError, onDone: () {
+      _addingStream = false;
+      completer.complete();
+    });
+    return completer.future;
+  }
+
+  @override
+  Future<void> close() {
+    if (_addingStream) throw StateError('StreamSink is bound to a stream');
+    return super.close();
+  }
+
+  @override
+  void _checkCanAddEvent() {
+    super._checkCanAddEvent();
+    if (_addingStream) throw StateError('StreamSink is bound to a stream');
+  }
+}
+
+/// An abstract class that implements `dart:io`'s [IOSink]'s API in terms of
+/// [onData], [onError], [onClose], and [onFlush] methods.
+///
+/// Because [IOSink] is defined in `dart:io`, this can't officially implement
+/// it. However, it's designed to match its API exactly so that subclasses can
+/// implement [IOSink] without any additional modifications.
+///
+/// This takes care of ensuring that events can't be added after [close] is
+/// called or during a call to [onStream].
+abstract class IOSinkBase extends StreamSinkBase<List<int>> {
```

Comment:
Not sure this class belongs in `package:async` *because* we can't depend on `dart:io`. I'd create a `package:io` instead if anything.

Reply:
I originally tried adding this to `package:io`, but it wants to share so much implementation with `StreamSinkBase` it would have produced a ton of duplication.

Also, I don't think `IOSink` belongs in `dart:io` in the first place, since nothing about its API requires or implies access to system IO. See https://github.com/dart-lang/sdk/issues/17293.